From 8b5b944fa9867d8aa73fefc7ad77cb17fdf6a9b5 Mon Sep 17 00:00:00 2001
From: Simon Gaiser <simon@invisiblethingslab.com>
Date: Wed, 24 Jan 2018 04:47:00 +0100
Subject: [PATCH] libxl: Add basic support for type=pvh

A bunch of stuff not needed for Qubes is missing (video, input, sound?,
cpu capabilities handling).
---
 docs/schemas/domaincommon.rng  |  2 --
 src/conf/domain_conf.c         |  7 ++++---
 src/conf/domain_conf.h         |  2 +-
 src/libxl/libxl_capabilities.c | 22 ++++++++++++++++++++++
 src/libxl/libxl_conf.c         | 34 ++++++++++++++++++++++++++++++----
 src/xenconfig/xen_common.c     |  8 ++++++++
 6 files changed, 65 insertions(+), 10 deletions(-)

diff --git a/docs/schemas/domaincommon.rng b/docs/schemas/domaincommon.rng
index de5955b9ac..ce25257769 100644
--- a/docs/schemas/domaincommon.rng
+++ b/docs/schemas/domaincommon.rng
@@ -3059,7 +3059,6 @@
                 <value>qemu</value>
                 <value>stubdom</value>
                 <value>stubdom-linux</value>
-                <value>none</value>
               </choice>
             </attribute>
           </optional>
@@ -3081,7 +3080,6 @@
               <value>qemu</value>
               <value>stubdom</value>
               <value>stubdom-linux</value>
-              <value>none</value>
             </choice>
           </attribute>
           <optional>
diff --git a/src/conf/domain_conf.c b/src/conf/domain_conf.c
index 5503e19314..1454eab98b 100644
--- a/src/conf/domain_conf.c
+++ b/src/conf/domain_conf.c
@@ -126,7 +126,8 @@ VIR_ENUM_IMPL(virDomainOS, VIR_DOMAIN_OSTYPE_LAST,
               "xen",
               "linux",
               "exe",
-              "uml")
+              "uml",
+              "pvh")
 
 VIR_ENUM_IMPL(virDomainBoot, VIR_DOMAIN_BOOT_LAST,
               "fd",
@@ -860,8 +861,7 @@ VIR_ENUM_IMPL(virDomainDiskTray, VIR_DOMAIN_DISK_TRAY_LAST,
 VIR_ENUM_IMPL(virDomainEmulatorType, VIR_DOMAIN_EMULATOR_TYPE_LAST,
               "qemu",
               "stubdom",
-              "stubdom-linux",
-              "none");
+              "stubdom-linux");
 
 VIR_ENUM_IMPL(virDomainRNGModel,
               VIR_DOMAIN_RNG_MODEL_LAST,
@@ -18838,6 +18838,7 @@ virDomainDefParseBootOptions(virDomainDefPtr def,
 
     if (def->os.type == VIR_DOMAIN_OSTYPE_XEN ||
         def->os.type == VIR_DOMAIN_OSTYPE_HVM ||
+        def->os.type == VIR_DOMAIN_OSTYPE_PVH ||
         def->os.type == VIR_DOMAIN_OSTYPE_UML) {
         xmlNodePtr loader_node;
 
diff --git a/src/conf/domain_conf.h b/src/conf/domain_conf.h
index 6888e32071..f528134a7d 100644
--- a/src/conf/domain_conf.h
+++ b/src/conf/domain_conf.h
@@ -255,6 +255,7 @@ typedef enum {
     VIR_DOMAIN_OSTYPE_LINUX,
     VIR_DOMAIN_OSTYPE_EXE,
     VIR_DOMAIN_OSTYPE_UML,
+    VIR_DOMAIN_OSTYPE_PVH,
 
     VIR_DOMAIN_OSTYPE_LAST
 } virDomainOSType;
@@ -2102,7 +2103,6 @@ typedef enum {
     VIR_DOMAIN_EMULATOR_TYPE_DEFAULT,
     VIR_DOMAIN_EMULATOR_TYPE_STUBDOM,
     VIR_DOMAIN_EMULATOR_TYPE_STUBDOM_LINUX,
-    VIR_DOMAIN_EMULATOR_TYPE_NONE,
 
     VIR_DOMAIN_EMULATOR_TYPE_LAST
 } virDomainEmulatorType;
diff --git a/src/libxl/libxl_capabilities.c b/src/libxl/libxl_capabilities.c
index 18596c7c72..3bc0b2f06f 100644
--- a/src/libxl/libxl_capabilities.c
+++ b/src/libxl/libxl_capabilities.c
@@ -564,6 +564,28 @@ libxlCapsInitGuests(libxl_ctx *ctx, virCapsPtr caps)
                                                1) == NULL)
                 return -1;
         }
+
+        if (guest_archs[i].hvm) {
+            if ((machines = virCapabilitiesAllocMachines(xen_machines, 1)) == NULL)
+                return -1;
+
+            if ((guest = virCapabilitiesAddGuest(caps,
+                                                 VIR_DOMAIN_OSTYPE_PVH,
+                                                 guest_archs[i].arch,
+                                                 LIBXL_EXECBIN_DIR "/qemu-system-i386",
+                                                 NULL,
+                                                 1,
+                                                 machines)) == NULL) {
+                virCapabilitiesFreeMachines(machines, 1);
+                return -1;
+            }
+            machines = NULL;
+
+            if (virCapabilitiesAddGuestFeature(guest, "apic",
+                                               1,
+                                               0) == NULL)
+                return -1;
+        }
     }
 
     return 0;
diff --git a/src/libxl/libxl_conf.c b/src/libxl/libxl_conf.c
index db2bea040f..4439ff8079 100644
--- a/src/libxl/libxl_conf.c
+++ b/src/libxl/libxl_conf.c
@@ -148,6 +148,8 @@ libxlMakeDomCreateInfo(libxl_ctx *ctx,
         case VIR_TRISTATE_SWITCH_LAST:
             break;
         }
+    } else if (def->os.type == VIR_DOMAIN_OSTYPE_PVH) {
+        c_info->type = LIBXL_DOMAIN_TYPE_PVH;
     } else {
         c_info->type = LIBXL_DOMAIN_TYPE_PV;
     }
@@ -277,6 +279,7 @@ libxlMakeDomBuildInfo(virDomainDefPtr def,
     libxl_ctx *ctx = cfg->ctx;
     libxl_domain_build_info *b_info = &d_config->b_info;
     int hvm = def->os.type == VIR_DOMAIN_OSTYPE_HVM;
+    bool pvh = def->os.type == VIR_DOMAIN_OSTYPE_PVH;
     size_t i;
     size_t nusbdevice = 0;
 
@@ -284,6 +287,8 @@ libxlMakeDomBuildInfo(virDomainDefPtr def,
 
     if (hvm)
         libxl_domain_build_info_init_type(b_info, LIBXL_DOMAIN_TYPE_HVM);
+    else if (pvh)
+        libxl_domain_build_info_init_type(b_info, LIBXL_DOMAIN_TYPE_PVH);
     else
         libxl_domain_build_info_init_type(b_info, LIBXL_DOMAIN_TYPE_PV);
 
@@ -549,8 +554,6 @@ libxlMakeDomBuildInfo(virDomainDefPtr def,
                 b_info->device_model_version = LIBXL_DEVICE_MODEL_VERSION_QEMU_XEN_TRADITIONAL;
             else if (def->emulator_type == VIR_DOMAIN_EMULATOR_TYPE_STUBDOM_LINUX)
                 b_info->device_model_version = LIBXL_DEVICE_MODEL_VERSION_QEMU_XEN;
-            else if (def->emulator_type == VIR_DOMAIN_EMULATOR_TYPE_NONE)
-                b_info->device_model_version = LIBXL_DEVICE_MODEL_VERSION_NONE;
         }
 
         /* In case of stubdom there will be two qemu instances:
@@ -662,6 +665,29 @@ libxlMakeDomBuildInfo(virDomainDefPtr def,
             return -1;
         }
 #endif
+    } else if (pvh) {
+        libxl_defbool_set(&b_info->u.pvh.apic,
+                          def->features[VIR_DOMAIN_FEATURE_APIC] ==
+                          VIR_TRISTATE_SWITCH_ON);
+
+        libxl_defbool_set(&b_info->u.pvh.nested_hvm, false);
+
+        if (def->os.bootloader) {
+            if (VIR_STRDUP(b_info->u.pvh.bootloader, def->os.bootloader) < 0)
+                return -1;
+        }
+        if (def->os.bootloaderArgs) {
+            if (!(b_info->u.pvh.bootloader_args =
+                  virStringSplit(def->os.bootloaderArgs, " \t\n", 0)))
+                return -1;
+        }
+
+        if (VIR_STRDUP(b_info->cmdline, def->os.cmdline) < 0)
+            return -1;
+        if (VIR_STRDUP(b_info->kernel, def->os.kernel) < 0)
+            return -1;
+        if (VIR_STRDUP(b_info->ramdisk, def->os.initrd) < 0)
+            return -1;
     } else {
         /*
          * For compatibility with the legacy xen toolstack, default to pygrub
@@ -1257,11 +1283,11 @@ libxlMakeNic(virDomainDefPtr def,
      * hvm guest").
      */
     if (l_nic->model) {
-        if (def->os.type == VIR_DOMAIN_OSTYPE_XEN &&
+        if (def->os.type != VIR_DOMAIN_OSTYPE_HVM &&
             STRNEQ(l_nic->model, "netfront")) {
             virReportError(VIR_ERR_CONFIG_UNSUPPORTED, "%s",
                            _("only model 'netfront' is supported for "
-                             "Xen PV domains"));
+                             "Xen PV(H) domains"));
             return -1;
         }
         if (VIR_STRDUP(x_nic->model, l_nic->model) < 0)
diff --git a/src/xenconfig/xen_common.c b/src/xenconfig/xen_common.c
index 16e08007e4..7ffca30833 100644
--- a/src/xenconfig/xen_common.c
+++ b/src/xenconfig/xen_common.c
@@ -592,6 +592,11 @@ xenParseCPUFeatures(virConfPtr conf,
 
             def->clock.timers[def->clock.ntimers - 1] = timer;
         }
+    } else if (def->os.type == VIR_DOMAIN_OSTYPE_PVH) {
+        if (xenConfigGetBool(conf, "apic", &val, 1) < 0)
+            return -1;
+        else if (val)
+            def->features[VIR_DOMAIN_FEATURE_APIC] = VIR_TRISTATE_SWITCH_ON;
     } else {
         if (xenConfigGetBool(conf, "e820_host", &val, 0) < 0) {
             return -1;
@@ -2019,6 +2024,9 @@ xenDomainDefAddImplicitInputDevice(virDomainDefPtr def)
 {
     virDomainInputBus implicitInputBus = VIR_DOMAIN_INPUT_BUS_XEN;
 
+    if (def->os.type == VIR_DOMAIN_OSTYPE_PVH)
+        return 0;
+
     if (def->os.type == VIR_DOMAIN_OSTYPE_HVM)
         implicitInputBus = VIR_DOMAIN_INPUT_BUS_PS2;
 
-- 
2.13.6

